### PR TITLE
Foreach element

### PR DESCRIPTION
**Pull request by teknoraver**
Add a utility macro to iterate an array of fixed, known size.

### CODE REVIEW

**Comment by YHNdnzj** [comment]
I prefer to name this `FOREACH_ELEMENT`, as mentioned in https://github.com/systemd/systemd/pull/31072#issuecomment-1908441353.

**Review by yuwata — CHANGES_REQUESTED** [review]
Yeah, `FOREACH_ELEMENT()` please.

File: src/shared/gpt.c [hunk 1]
```diff
@@ -192,7 +192,7 @@ const GptPartitionType gpt_partition_type_table[] = {
 
 static const GptPartitionType *gpt_partition_type_find_by_uuid(sd_id128_t id) {
 
-        FOREACH_ARRAY(t, gpt_partition_type_table, ELEMENTSOF(gpt_partition_type_table) - 1)
```
Comment (YHNdnzj) [inline src/shared/gpt.c]: This should not be switched over.

File: src/shared/gpt.c [hunk 2]
```diff
@@ -230,7 +230,7 @@ int gpt_partition_type_from_string(const char *s, GptPartitionType *ret) {
 
         assert(s);
 
-        FOREACH_ARRAY(t, gpt_partition_type_table, ELEMENTSOF(gpt_partition_type_table) - 1)
+        FOREACH_ELEMENT(t, gpt_partition_type_table)
```
Comment (YHNdnzj) [inline src/shared/gpt.c]: ditto

File: src/shared/gpt.c [hunk 3]
```diff
@@ -253,7 +253,7 @@ int gpt_partition_type_from_string(const char *s, GptPartitionType *ret) {
 GptPartitionType gpt_partition_type_override_architecture(GptPartitionType type, Architecture arch) {
         assert(arch >= 0);
 
-        FOREACH_ARRAY(t, gpt_partition_type_table, ELEMENTSOF(gpt_partition_type_table) - 1)
```
Comment (YHNdnzj) [inline src/shared/gpt.c]: and here

**Review by YHNdnzj — CHANGES_REQUESTED** [review]

File: src/shared/gpt.c (hunk 1 above)
Comment (yuwata) [reply src/shared/gpt.c]: Or, drop the last empty element in the table, if that works.

Comment (YHNdnzj) [reply src/shared/gpt.c]: Hmm, the table is exported, and external users need the sentinel I presume.

Comment (yuwata) [reply src/shared/gpt.c]: Exported? In what sense?

Comment (YHNdnzj) [reply src/shared/gpt.c]: `gpt_partition_type_table` is used outside of `shared/gpt.c`. ELEMENTSOF/sizeof can't be used there.

Comment (yuwata) [reply src/shared/gpt.c]: Indeed...

Comment (teknoraver) [reply src/shared/gpt.c]: fixed

**Review by YHNdnzj — APPROVED** [review]

**Review by yuwata — APPROVED** [review]

**Pull request merged**